flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushPartitionIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,179 @@
+		RelDataType inputFieldType = filter.getInput().getRowType();
+		List<String> inputFieldName = inputFieldType.getFieldNames();
+
+		List<String> partitionedFieldNames = tableSourceTable.catalogTable().getPartitionKeys();
+
+		RelBuilder relBuilder = call.builder();
+		RexBuilder rexBuilder = relBuilder.getRexBuilder();
+
+		Tuple2<Seq<RexNode>, Seq<RexNode>> predicate = RexNodeExtractor.extractPartitionPredicateList(
+			filter.getCondition(),
+			FlinkRelOptUtil.getMaxCnfNodeCount(scan),
+			inputFieldName.toArray(new String[inputFieldName.size()]),
+			rexBuilder,
+			partitionedFieldNames.toArray(new String[partitionedFieldNames.size()])
+			);
+
+		RexNode partitionPredicate = RexUtil.composeConjunction(rexBuilder, JavaConversions.seqAsJavaList(predicate._1));
+
+		if (partitionPredicate.isAlwaysTrue()){
+			return;
+		}
+
+		List<LogicalType> partitionFieldType = partitionedFieldNames.stream().map(name -> {
+			int index = inputFieldName.indexOf(name);
+			if (index < 0) {
+				throw new RuntimeException(String.format("Partitioned key '%s' isn't found in input columns. " +
+					"Validator should have checked that.", name));
+			}
+			return inputFieldType.getFieldList().get(index).getType(); })
+			.map(FlinkTypeFactory::toLogicalType).collect(Collectors.toList());
+
+		// checked in matches(call) and listPartitions() is not empty.
+		List<Map<String, String>> allPartitions = ((SupportsPartitionPushDown) dynamicTableSource).listPartitions().get();
+		RexNode finalPartitionPredicate = adjustPartitionPredicate(inputFieldName, partitionedFieldNames, partitionPredicate);
+		// get partitions
+		List<Map<String, String>> remainingPartitions = PartitionPruner.prunePartitions(
+			((FlinkContext) context).getTableConfig(),
+			partitionedFieldNames.toArray(new String[partitionedFieldNames.size()]),
+			partitionFieldType.toArray(new LogicalType[partitionFieldType.size()]),
+			allPartitions,
+			finalPartitionPredicate
+			);
+		((SupportsPartitionPushDown) dynamicTableSource).applyPartitions(remainingPartitions);
+
+		// build new table scan
+		FlinkStatistic statistic = tableSourceTable.getStatistic();
+		String extraDigest = "source: [partitions=" +
+			String.join(", ", ((SupportsPartitionPushDown) dynamicTableSource).listPartitions()
+				.get()
+				.stream()
+				.map(partition -> partition.toString())
+				.collect(Collectors.toList())
+				.toArray(new String[1])) +
+			"]";
+		TableSourceTable newTableSourceTable = tableSourceTable.copy(dynamicTableSource, statistic, new String[]{extraDigest});
+		LogicalTableScan newScan = new LogicalTableScan(
+			scan.getCluster(), scan.getTraitSet(), scan.getHints(), newTableSourceTable);
+
+		RexNode nonPartitionPredicate = RexUtil.composeConjunction(rexBuilder, JavaConversions.seqAsJavaList(predicate._2()));
+		if (nonPartitionPredicate.isAlwaysTrue()) {
+			call.transformTo(newScan);
+		} else {
+			call.transformTo(filter.copy(filter.getTraitSet(), newScan, nonPartitionPredicate));
```

Comment:
move `filter.copy(filter.getTraitSet(), newScan, nonPartitionPredicate)` into a new line for more easy debugging